core/src/main/scala/kafka/tools/StorageTool.scala

```diff
@@ -128,6 +152,108 @@ object StorageTool extends Logging {
       .getOrElse(defaultValue)
   }
 
+  def getUserScramCredentialRecord(mechanism: String,
```

Comment:
This is a lot of parsing logic (which is fine, that's what we expected.)

I think this would be better as Java, and in a separate file. After all, we will eventually be rewriting this tool in Java not too far down the line.

You could probably put it in `./metadata/src/main/java/org/apache/kafka/metadata/util` as its own file

You will also need a lot of unit tests for this, including ones where parsing fails.

Reply:
Lots of unit tests were added in StorageToolTest.scala . I made sure to cover all the required argument combinations.

Reply:
I plan to do this later when I have time. Maybe I'll write it in Rust. :)

Reply:
There are a lot of contributors who are rewriting tools in Java right now. You can take a look at the `git log` of the last month or two:

    KAFKA-14578: Move ConsumerPerformance to tools (#13215)
    KAFKA-14590: Move DelegationTokenCommand to tools (#13172)
    KAFKA-14582: Move JmxTool to tools (#13136)
    KAFKA-14575: Move ClusterTool to tools module (#13080)
    MINOR: Move MetadataQuorumCommand from `core` to `tools` (#12951)
    ... etc ...

These rewrites aren't a "maybe someday" kind of thing; they are happening right now.

Anyway, like I said earlier, we don't have to do that in this PR :) None of this will be hard to port to Java when the time comes.